abdiff/core/collate_ab_transforms.py

```diff
@@ -0,0 +1,220 @@
+import itertools
+import json
+import logging
+import re
+import tempfile
+from collections.abc import Generator
+from pathlib import Path
+
+import duckdb
+import ijson
+import pyarrow as pa
+
+from abdiff.core.utils import write_to_dataset
+
+logger = logging.getLogger(__name__)
+
+READ_BATCH_SIZE = 1_000
+TRANSFORMED_DATASET_SCHEMA = pa.schema(
+    [
+        pa.field("timdex_record_id", pa.string()),
+        pa.field("source", pa.string()),
+        pa.field("record", pa.binary()),
+        pa.field("version", pa.string()),
+        pa.field("transformed_file_name", pa.string()),
+    ]
+)
+JOINED_DATASET_SCHEMA = pa.schema(
+    [
+        pa.field("timdex_record_id", pa.string()),
+        pa.field("source", pa.string()),
+        pa.field("record_a", pa.binary()),
+        pa.field("record_b", pa.binary()),
+    ]
+)
+
+
+def collate_ab_transforms(
+    run_directory: str, transformed_files: tuple[list[str], ...]
+) -> str:
+    """Collates A/B transformed files into a Parquet dataset.
+
+    This process can be summarized into two (2) important steps:
+        1. Write all transformed JSON records into a temporary Parquet dataset
+           partitioned by the transformed file name.
+        2. For every transformed file, use DuckDB to join A/B Parquet tables
+           using the TIMDEX record ID and write joined records to a Parquet dataset.
+
+    This function (and its subfunctions) uses DuckDB, generators, and batching to
+    write records to Parquet datasets in a memory-efficient manner.
+    """
+    transformed_dataset_path = tempfile.TemporaryDirectory()
+    collated_dataset_path = str(Path(run_directory) / "collated")
+
+    transformed_written_files = write_to_dataset(
+        get_transformed_batches_iter(run_directory, transformed_files),
+        schema=TRANSFORMED_DATASET_SCHEMA,
+        base_dir=transformed_dataset_path.name,
+        partition_columns=["transformed_file_name"],
+    )
+    logger.info(
+        f"Wrote {len(transformed_written_files)} parquet file(s) to transformed dataset"
+    )
+
+    joined_written_files = write_to_dataset(
+        get_joined_batches_iter(transformed_dataset_path.name),
+        base_dir=collated_dataset_path,
+        schema=JOINED_DATASET_SCHEMA,
+    )
+    logger.info(f"Wrote {len(joined_written_files)} parquet file(s) to collated dataset")
+    return collated_dataset_path
+
+
+def yield_records(transformed_file: str | Path) -> Generator:
```

Comment:
I think the name is misleading, maybe `yield_data_from_records`? Otherwise, I assume full records as the output. Also, I would expand the type hint to `Generator[dict[str,str]` to further clarify the output

Reply:
Hmm, what do you think about renaming to `get_transformed_records_iter` (to be consistent with the other `Generator` functions in `collate_ab_transforms.py`)? Also curious to get @ghukill 's as he helped with this function. 🤔 

As a reminder, this is what the function _yields_:

```python
 yield {
   "timdex_record_id": record["timdex_record_id"],
   "source": record["source"],
   "record": json.dumps(record).encode(),
   "version": version,
   "transformed_file_name": transformed_file_name,
}
```
It returns the TIMDEX record _serialized to a JSON string_ + other vars to help with the dataset creation downstream...

Reply:
My apologies, I was completely misreading this so please ignore my objections!

Reply:
`get_transformed_records_iter()` makes sense to me!

I think `yield_data_from_records()` was kind of a stubbed / half-baked name, apologies.  I like this better.